Parsing of <body link vlink alink> is not per spec

rniwa opened the pull request:
#### c521422a249b8f1153631376f5179ce8b81700eb
<pre>
Parsing of &lt;body link vlink alink&gt; is not per spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=156849">https://bugs.webkit.org/show_bug.cgi?id=156849</a>

Reviewed by Darin Adler.

Use the rules for parsing a legacy color value to parse link, vlink, and alink content attributes
on body element: <a href="https://html.spec.whatwg.org/#rules-for-parsing-a-legacy-colour-value">https://html.spec.whatwg.org/#rules-for-parsing-a-legacy-colour-value</a>

Based on <a href="https://crrev.com/ac6c480c0c078999a785de7ef1e10c6c0c928c75">https://crrev.com/ac6c480c0c078999a785de7ef1e10c6c0c928c75</a>

* LayoutTests/fast/html/body-color-legacy-parsing-1-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-1.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-2-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-2.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-3-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-3.html: Added.
* LayoutTests/platform/ios/TestExpectations:

* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::parseAttribute): Fixed the bug.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::parseLegacyColorValue): Made this a protected member function.
(WebCore::HTMLElement::addHTMLColorToStyle):
* Source/WebCore/html/HTMLElement.h:

Canonical link: <a href="https://commits.webkit.org/252992@main">https://commits.webkit.org/252992@main</a>
</pre>

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -148,7 +148,8 @@ class HTMLElement : public StyledElement {
     void addHTMLNumberToStyle(MutableStyleProperties&, CSSPropertyID, StringView value);
 
-    void addHTMLColorToStyle(MutableStyleProperties&, CSSPropertyID, const String& color);
+    std::optional<SRGBA<uint8_t>> parseLegacyColorValue(StringView);
```
I suggest we make this a static member function.

rniwa replied:
That's a good point. Will do.

darinadler commented on Source/WebCore/html/HTMLBodyElement.cpp:
```diff
@@ -115,25 +115,24 @@ const AtomString& HTMLBodyElement::eventNameForWindowEventHandlerAttribute(const
+        auto parsedColor = parseLegacyColorValue(value);
+        if (name == linkAttr) {
+            if (parsedColor)
+                document().setLinkColor(*parsedColor);
```
We should consider refactoring later and just have set functions on the document class that take an optional rather than pairs of set and reset functions. Also noting that since `Color` has an invalid value we could consider using that instead of `std::optional`. I guess that’s what `WTF::Markable` is for?

rniwa replied:
Yeah, that's probably a good refactoring to do afterwards.

rniwa committed:
Parsing of <body link vlink alink> is not per spec
https://bugs.webkit.org/show_bug.cgi?id=156849

Reviewed by Darin Adler.

Use the rules for parsing a legacy color value to parse link, vlink, and alink content attributes
on body element: https://html.spec.whatwg.org/#rules-for-parsing-a-legacy-colour-value

Based on https://crrev.com/ac6c480c0c078999a785de7ef1e10c6c0c928c75

* LayoutTests/fast/html/body-color-legacy-parsing-1-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-1.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-2-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-2.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-3-expected.html: Added.
* LayoutTests/fast/html/body-color-legacy-parsing-3.html: Added.
* LayoutTests/platform/ios/TestExpectations:

* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::parseAttribute): Fixed the bug.
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::parseLegacyColorValue): Made this a protected member function.
(WebCore::HTMLElement::addHTMLColorToStyle):
* Source/WebCore/html/HTMLElement.h:

Canonical link: https://commits.webkit.org/252992@main
